src/builtins.c

```diff
@@ -602,6 +602,14 @@ JL_CALLABLE(jl_f__apply)
             jl_value_t *next = jl_apply_generic(jl_iterate_func, args, 1);
             while (next != jl_nothing) {
                 roots[stackalloc] = next;
+                if (!jl_is_tuple(next) || jl_nfields(next) < 2) {
```

Comment:
I don’t think that for loops normally assert the type

Reply:
Would you prefer a BoundsError?

Reply:
If we just call `jl_get_nth_field_checked` below it should match what `for` lowering usually does.

Reply:
Yes, that's fine. I figured I'd hoist the error and be a bit more informative that it's happening inside the _apply call (since it can be a bit confusing), but I'm happy to just do that instead.